[Tan How Suen] iP

HowSuen opened the pull request:
# Duke
> "I am Duke." - Duke

Duke is an interactive chat-bot that helps you manage your tasks!

### Duke is,
- text-based
- easy to learn
- very friendly!

Best of all, it is ~~only $69,420!~~ **FREE!**

### All you need to do is,
1. Download Duke from this [link](https://github.com/HowSuen/ip/releases)
2. Run ```java -jar duke.jar```
3. Add your tasks
4. Let it manage your tasks for you 😉 

## Features:
- [X] Add ToDo, Deadline and Event tasks
- [X] Date/time for Deadlines and Events (dd MM yyyy HH:mm)
- [X] Remove tasks
- [X] Display all your tasks
- [X] Mark your tasks as done or not done
- [X] Search for tasks
- [ ] Reminder for tasks _(Coming soon)_
--------------
Run the `main` method:
```java
public static void main(String[] args) {
        Duke duke = new Duke();
        duke.run();
}
```
----------------
Have fun with Duke! 🤖

Commit:
Add Gradle support

HowSuen committed:
Level 1. Greet, Echo, Exit

HowSuen committed:
Added TasksList with Add and List features

HowSuen committed:
Added Tasks, mark and unmark tasks functions

HowSuen committed:
Added ToDo, Deadline, Event

HowSuen committed:
Added Automated Text UI Testing

HowSuen committed:
Added DukeException for incorrect user inputs

HowSuen committed:
Added delete-tasks feature

HowSuen committed:
Edited variable name in Task

HowSuen committed:
Added TaskType enums

HowSuen committed:
Add Storage to store tasks in hard disk

HowSuen committed:
Add DateParser to accept dates and times

HowSuen committed:
Merge branch 'branch-Level-7'

HowSuen committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/TasksList.java

HowSuen committed:
Add Ui, Parser and Commands

HowSuen committed:
Add Packages and edit tests

HowSuen committed:
Fix the name of Duke in README

HowSuen committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

HowSuen committed:
Add Gradle support, JUnit Tests and checkstyle

HowSuen committed:
Edit build.gradle

HowSuen committed:
Add JavaDoc comments

HowSuen committed:
Fix code to comply with coding standard

HowSuen committed:
Add Find command to search for task

HowSuen committed:
Merge branch 'branch-A-JavaDoc'

HowSuen committed:
Merge branch 'branch-A-CodingStandard'

HowSuen committed:
Merge branch 'branch-Level-9'

HowSuen committed:
Fix style in tests and add find to runtest

HowSuen committed:
Add JavaDoc comments

HowSuen committed:
Add HelpCommand to list all commands

HowSuen committed:
Edit error messages and check style

HowSuen committed:
Merge branch 'branch-A-Checkstyle'

HowSuen committed:
Configure build.gradle to set up JavaFX

HowSuen committed:
Add GUI to Duke with JavaFX and refactor TasksList to TaskList

HowSuen committed:
Merge branch 'branch-Level-10'

HoJunHao2000 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,52 @@
+    public String toString() {
+        return "Duke";
+    }
+
```
Missing JavaDocs for public method getResponse().

HoJunHao2000 commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,40 @@
+public class Main extends Application {
+    private static Stage stage;
+    private Duke duke = new Duke();
+
```
Missing JavaDocs.

HoJunHao2000 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,172 @@
+    public void addTaskToSave(Task task) throws DukeException {
+        try {
+            List<String> history = Files.readAllLines(this.path);
+            history.add(task.toCommand());
```
Very smart and unique way of adding/removing tasks from list instead of passing copying directly from the tasklist.

HoJunHao2000 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+     * @param s The command word (if any) from the user's input.
+     * @return true if the command to delete tasks is valid, false otherwise.
+     */
+    public static boolean isCommand(String s) {
```
you could perhaps consider moving isCommand into the abstract Command class and creating an isEnd variable in the command superclass. When executing a delete command, change the variable and check from there.

HoJunHao2000 reviewed (commented):
Overall the code was very well organised. JavaDocs were done well. Classes, methods, and variables were all appropriately named!

HowSuen committed:
Add JavaDoc comments

HowSuen committed:
Add greeting message for GUI

Neethesh26 commented on src/main/java/duke/task/TaskType.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Represents the single alphabet indicating the type of the Task.
+ */
+public enum TaskType {
+    T, // ToDo
+    D, // Deadline
+    E // Event
```
Good use of enum!

Neethesh26 commented on src/main/java/duke/ui/MainWindow.java:
```diff
@@ -0,0 +1,72 @@
+ * Controller for MainWindow. Provides the layout for the other controls.
+ */
+public class MainWindow extends AnchorPane {
+    @FXML
+    private ScrollPane scrollPane;
+    @FXML
+    private VBox dialogContainer;
+    @FXML
+    private TextField userInput;
+    @FXML
+    private Button sendButton;
+
+    private Duke duke;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/User.png"));
+    private Image dukeImage = new Image(this.getClass().getResourceAsStream("/images/Duke.png"));
+
+    /**
```
Good job in splitting up the logical units for better readability

Neethesh26 reviewed (commented):
Generally a really great job in adhering to code quality! Keep up the good work :))

HowSuen committed:
Edit Duke's responses and rename methods

Removed "Duke: " strings from Duke's responses.
Edited DukeException and error messages by removing
"☹ OOPS!!! Duke got an error!\n" because emoji is not visible on
JavaFX GUI.
Renamed methods in DateParser to include verbs.

HowSuen committed:
Add Assertions to TaskList and Storage

Use `assert` feature to document important assumptions that should
hold at various points in the code.

Assertions added to include checking of the right command-word and
the length of the command String array in the functions of TaskList.

Assertions added to ensure the data stored in the save file is valid
before loading it into Duke.

HowSuen committed:
Improve code quality

Removed unused and unnecessary code, specifically the `showGreeting`
and `showMessage` methods from Ui and Duke.

Added comments to explain date/time formats.

Changed the checks on the AddCommand to avoid complicated
expressions.

HowSuen committed:
Add use of Streams feature

Added Streams to create String and StringBuilder of the TaskList
and replaced use of loops in `Storage` and `TaskList`

HowSuen committed:
Merge pull request #1 from HowSuen/branch-A-Assertions

Add Assertions to TaskList and Storage

HowSuen committed:
Merge pull request #2 from HowSuen/branch-A-CodeQuality

Improve code quality

HowSuen committed:
Merge pull request #3 from HowSuen/branch-A-Streams

Add use of Streams feature

HowSuen committed:
Merge branch 'master' into branch-A-Streams

HowSuen committed:
Merge pull request #4 from HowSuen/branch-A-Streams

Merge branch 'master' into branch-A-Streams

HowSuen committed:
Merge branch 'master' of https://github.com/HowSuen/ip

* 'master' of https://github.com/HowSuen/ip:

HowSuen committed:
Add editD and editT commands to edit a task

`editD <task_id> <new_description>` to edit the description of the
task with the given id in the list.
`editT <task_id> <new_date/time>` to edit the date/time of the task
with the given id in the list.

Updated the list of instructions in the `help` command.

HowSuen committed:
Merge branch 'branch-C-Update'

HowSuen committed:
Set up Github Actions Continuous Integration (CI)

HowSuen committed:
Fix coding standard

HowSuen committed:
Improve GUI design

HowSuen committed:
Refactor Duke to Raiden

Changed name of bot from Duke to Raiden.
Changed duke folder to raiden.

HowSuen committed:
Refactor DukeException to RaidenException

Renamed DukeException to RaidenException and edited Raiden's messages.

HowSuen committed:
Merge branch 'branch-A-Personality'

HowSuen committed:
Add speech bubbles and colors for textfield

HowSuen committed:
Adjust color for button and textfield

HowSuen committed:
Change font size to 12

HowSuen committed:
Add User guide in README

Updated find command to search for keywords regardless of lower or
upper case.

HowSuen committed:
Update the README in docs

HowSuen committed:
Add responsiveness to GUI

HowSuen committed:
Change screenshots and add new greet image

HowSuen committed:
Fix resizing of VBox and refactor Duke methods

HowSuen committed:
Update README and change font of Send button

HowSuen committed:
Ammend README

HowSuen committed:
Add config file for Github page

HowSuen committed:
Fix emoji in README

HowSuen committed:
Amend README

HowSuen committed:
Remove javafx from build.gradle file

Possible cause for failing automated smoke testing of iP JAR file.
